File: rules.py
from moves import Move, PlaceWall, PlacePawn, MovePawn
from board import Field, Pawn
from path import findValidPath

class Referee():
    
    def __init__(self):
        self.active = True
        
    def toggleActive(self):
        self.active = not self.active
        
    def isActive(self):
        return self.active
    
    def checkLegalMove(self, move: Move, board_state: dict) -> bool:
        if not self.active:
            return True
        match move:
            case PlaceWall():
                return self.checkWallPlace(move.coordinates, move.direction, board_state)
            case PlacePawn():
                return self.checkPawnPlace(move.coordinates, move.player, board_state)
            case MovePawn():
                return self.checkPawnMove(move.start_coordinates, move.end_coordinates, board_state)
            case _:
                raise ValueError('Invalid move')
            
    def checkWallPlace(self, coordinates: tuple[int, int], direction: str, board_state: dict):
        # Check if wall is already placed
        if board_state['fields'][coordinates[0], coordinates[1]].getWall(direction):
            return False
        # Check if wall is placed on the edge of the board
        if coordinates[0] == 0 and direction == 'W':
            return False
        if coordinates[0] == board_state['size'] - 1 and direction == 'E':
            return False
        if coordinates[1] == 0 and direction == 'N':
            return False
        if coordinates[1] == board_state['size'] - 1 and direction == 'S':
            return False
        # Check if wall placement is next to the previous pawn, moved by the same player
        pawn_positions = []
        previous_move = board_state['moves_list'][-1]
        if isinstance(previous_move, MovePawn):
            if previous_move.player == board_state['turn']:
                pawn_positions.append(previous_move.end_coordinates)
        else: # PlacePawn or PlaceWall
            if previous_move.player == board_state['turn']:
                return False
            else:
                active_pawns = board_state['pawns1'] if board_state['turn'] == 1 else board_state['pawns2']
                for pawn in active_pawns:
                    pawn_positions.append(pawn.getPosition())
        for pawn_coordinates in pawn_positions:
            if pawn_coordinates == coordinates:
                return True
            if pawn_coordinates == (coordinates[0] - 1, coordinates[1]) and direction == 'W':
                return True
            if pawn_coordinates == (coordinates[0] + 1, coordinates[1]) and direction == 'E':
                return True
            if pawn_coordinates == (coordinates[0], coordinates[1] - 1) and direction == 'N':
                return True
            if pawn_coordinates == (coordinates[0], coordinates[1] + 1) and direction == 'S':
                return True
        return False


    def checkPawnMove(self, start_coordinates: tuple[int, int], end_coordinates: tuple[int, int], board_state: dict):
        previous_move = board_state['moves_list'][-1]
        if previous_move.player == board_state['turn']:
            return False
        if not board_state['fields'][start_coordinates[0], start_coordinates[1]].getPawn().isActive():
            return False
        return findValidPath(start_coordinates, end_coordinates, board_state['fields'])
    
    def checkPawnPlace(self, coordinates: tuple[int, int], player: int, board_state: dict):
        player_pawns: list[Pawn] = board_state['pawns1'] if player == 1 else board_state['pawns2']
        active_pawns = [pawn for pawn in player_pawns if pawn.isActive()]
        
        if board_state['max_pawns'] - len(player_pawns) == 0:
            return False
        
        previous_move = board_state['moves_list'][-1]
        if isinstance(previous_move, PlacePawn):
            if previous_move.player == player:
                return False
        if isinstance(previous_move, MovePawn):
            return False
        for pawn in active_pawns:
            if findValidPath(pawn.getPosition(), coordinates, board_state['fields']):
                return True
        return False
    

File: rules.cpp
#include "gamerep.h"
#include "path.cpp"

/* Notes: direction is a transferred mask*/
bool checkWallPlace(char x, char y, unsigned char direction, field_t * boardState){
    field_t* field = boardState + x + 7 * y;
    // Check if wall is already placed
    if (*field & direction){
        return false;
    }
    // Check if wall is placed on the edge of the board
    if (y == 0 && direction == NORTH){
        return false;
    }
    if (y == 6 && direction == SOUTH){
        return false;
    }
    if (x == 6 && direction == EAST){
        return false;
    }
    if (x == 0 && direction == WEST) {
        return false;
    }
    // Check if wall placement is next to the previous pawn, moved by the same player
    /* Not needed for AI */
    // Check if field has pawn
    /* Not needed for AI */
    return true;
}


bool checkPawnMove(char x, char y, char u, char v, field_t* boardState){
    field_t* startField = boardState + x + 7*y;
    field_t* endField = boardState + u + 7*v;
    // Check if it is the correct turn
    /* Not needed for AI */
    // Check if the start field has an active pawn (For AI move generation this equals to checking if the start field is already assigned or not)
    if (*startField & ASSIGNED){
        return false;
    }
    // Check if the end field is occupied
    if (*endField & OCCUPIED){
        return false;
    }
    return findValidPath(x, y, u, v, boardState);
}


bool checkPawnPlace(char u, char v, char player, field_t* boardState){
    field_t *placedPawns, *iField;
    char playerPawns, x , y;
    // Select correct pawns
    if (player = 1){
        placedPawns = boardState + PAWNS1NUM;
        playerPawns = PLAYER1PAWN;
    }
    else {
        placedPawns = boardState + PAWNS2NUM;
        playerPawns = PLAYER2PAWN;
    }
    // Check if pawns left (hardcoded max pawns)
    if ((7 - (char)*placedPawns) <= 0){
        return false;
    }
    // Check if PlacePawn is done as only move of the correct turns player
    /* Not needed for AI */
    for (int i = 0; i < 49; i++){
        iField = boardState + i;
        if( !(*iField & ASSIGNED) && (*iField & playerPawns) ){
            x = i % 7;
            y = i / 7;
            if (findValidPath(x, y, u, v, boardState)){
                return true;
            }
        }
    }
    return false;
}